# Part One: Toronto boroughs

Install the required modules for web scraping

In [1]:
!pip install lxml html5lib beautifulsoup4

In [2]:
import pandas as pd

In [43]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
df = pd.read_html(url)

The first entry of the dataframe *df* is the table we are looking for.

In [44]:
df_toronto = df[0]

Drop the rows with a borough that is **Not assigned**.
<br>
Then sort the rows by **Postal Code**.

In [45]:
df_toronto = df_toronto[df_toronto['Borough'] != 'Not assigned']
df_toronto = df_toronto.sort_values(by = ['Postal Code'])
df_toronto.reset_index(drop = True, inplace=True)
df_toronto

Postal Code      Borough  \
0           M1B  Scarborough   
1           M1C  Scarborough   
2           M1E  Scarborough   
3           M1G  Scarborough   
4           M1H  Scarborough   
..          ...          ...   
98          M9N         York   
99          M9P    Etobicoke   
100         M9R    Etobicoke   
101         M9V    Etobicoke   
102         M9W    Etobicoke   

                                          Neighborhood  
0                                       Malvern, Rouge  
1               Rouge Hill, Port Union, Highland Creek  
2                    Guildwood, Morningside, West Hill  
3                                               Woburn  
4                                            Cedarbrae  
..                                                 ...  
98                                              Weston  
99                                           Westmount  
100  Kingsview Village, St. Phillips, Martin Grove ...  
101  South Steeles, Silverstone, Humbergate, Jamest...  
102                Northwest, West Humber - Clairville  

[103 rows x 3 columns]

Check the number of unique postal codes.

In [46]:
len(pd.unique(df_toronto['Postal Code']))

103

We have 103 unique postal codes, which is the total number of rows in df_toronto. No need to merge the neighborhoods for a given postal code.

Let's check is there is any boroughs that are not assigned. 

In [47]:
df_toronto[df_toronto['Borough'] == 'Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

Let's check is there is any neighborhoods that are not assigned. 

In [48]:
df_toronto[df_toronto['Neighborhood'] == 'Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

All boroughs and neighborhoods are assigned.

Let's check the shape of the dataframe.

In [49]:
df_toronto.shape

(103, 3)

# Part Two: Adding latitude and longitude

In [11]:
!pip install geocoder

In [50]:
import geocoder

Due to the unreliability of the geocoder, the following code wasn't fruitful in retrieving the latitudes and longitudes of the postal codes. 

In [51]:
#lat = []
#lng = []
#for postal_code in df_toronto['Postal Code']:
#    postal_code_latlng = None
#    while (postal_code_latlng is None):
#        geo = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
#        postal_code_latlng = geo.latlng
#    lat.append(postal_code_latlng[0])
#    lng.append(postal_code_latlng[1])
#df_toronto['Latitude'] = lat
#df_toronto['Longitude'] = lng

Load the csv file for the coordinates of the postal codes.

In [52]:
df_toronto_latlng = pd.read_csv(r'https://cocl.us/Geospatial_data')
df_toronto_latlng = df_toronto_latlng.sort_values(by = [ 'Postal Code'])

Combine the two dataframe.

In [53]:
df_toronto.head()

Postal Code      Borough                            Neighborhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

In [54]:
df_toronto_latlng.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [55]:
df_toronto = pd.concat([df_toronto, df_toronto_latlng[['Latitude', 'Longitude']]], axis = 1)

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [64]:
df_toronto.head(40)

Postal Code       Borough  \
0          M1B   Scarborough   
1          M1C   Scarborough   
2          M1E   Scarborough   
3          M1G   Scarborough   
4          M1H   Scarborough   
5          M1J   Scarborough   
6          M1K   Scarborough   
7          M1L   Scarborough   
8          M1M   Scarborough   
9          M1N   Scarborough   
10         M1P   Scarborough   
11         M1R   Scarborough   
12         M1S   Scarborough   
13         M1T   Scarborough   
14         M1V   Scarborough   
15         M1W   Scarborough   
16         M1X   Scarborough   
17         M2H    North York   
18         M2J    North York   
19         M2K    North York   
20         M2L    North York   
21         M2M    North York   
22         M2N    North York   
23         M2P    North York   
24         M2R    North York   
25         M3A    North York   
26         M3B    North York   
27         M3C    North York   
28         M3H    North York   
29         M3J    North York   
30         M3K    North York   
31         M3L    North York   
32         M3M    North York   
33         M3N    North York   
34         M4A    North York   
35         M4B     East York   
36         M4C     East York   
37         M4E  East Toronto   
38         M4G     East York   
39         M4H     East York   

                                         Neighborhood   Latitude  Longitude  
0                                      Malvern, Rouge  43.806686 -79.194353  
1              Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
2                   Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                              Woburn  43.770992 -79.216917  
4                                           Cedarbrae  43.773136 -79.239476  
5                                 Scarborough Village  43.744734 -79.239476  
6         Kennedy Park, Ionview, East Birchmount Park  43.727929 -79.262029  
7                     Golden Mile, Clairlea, Oakridge  43.711112 -79.284577  
8     Cliffside, Cliffcrest, Scarborough Village West  43.716316 -79.239476  
9                         Birch Cliff, Cliffside West  43.692657 -79.264848  
10  Dorset Park, Wexford Heights, Scarborough Town...  43.757410 -79.273304  
11                                  Wexford, Maryvale  43.750072 -79.295849  
12                                          Agincourt  43.794200 -79.262029  
13            Clarks Corners, Tam O'Shanter, Sullivan  43.781638 -79.304302  
14  Milliken, Agincourt North, Steeles East, L'Amo...  43.815252 -79.284577  
15                      Steeles West, L'Amoreaux West  43.799525 -79.318389  
16                                        Upper Rouge  43.836125 -79.205636  
17                                  Hillcrest Village  43.803762 -79.363452  
18                       Fairview, Henry Farm, Oriole  43.778517 -79.346556  
19                                    Bayview Village  43.786947 -79.385975  
20                           York Mills, Silver Hills  43.757490 -79.374714  
21                            Willowdale, Newtonbrook  43.789053 -79.408493  
22                        Willowdale, Willowdale East  43.770120 -79.408493  
23                                    York Mills West  43.752758 -79.400049  
24                        Willowdale, Willowdale West  43.782736 -79.442259  
25                                          Parkwoods  43.753259 -79.329656  
26                                          Don Mills  43.745906 -79.352188  
27                                          Don Mills  43.725900 -79.340923  
28    Bathurst Manor, Wilson Heights, Downsview North  43.754328 -79.442259  
29                    Northwood Park, York University  43.767980 -79.487262  
30                                          Downsview  43.737473 -79.464763  
31                                          Downsview  43.739015 -79.506944  
32                                          Downsview  43.728496 -79.495697  
33                                    

# Part Three: Clustering

Import the libraries needed for data analysis and visualisation. 

In [34]:
import numpy as np # library to handle data in a vectorized manner

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

Solving environment: done

## Package Plan ##

  environment location: /opt/anaconda3

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geopy-2.0.0                |     pyh9f0ad1d_0          63 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          63 KB

The following packages will be UPDATED:

  geopy                                 1.22.0-pyh9f0ad1d_0 --> 2.0.0-pyh9f0ad1d_0



geopy-2.0.0          | 63 KB     | ##################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: done

## Package Plan ##

  environment location: /opt/anaconda3

  added / updated specs:
    - folium=0.5.0


The followi

First we show the neighborhoods in Toronto using their geographical coordinates.

In [62]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[43.7032, -79.3832], zoom_start=10.5)

# add markers to map
for lat, lng, label in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [86]:
CLIENT_ID = 'xxx' 
CLIENT_SECRET = 'xxx' 
VERSION = '20180605' # Foursquare API version
LIMIT = 100

Next, we define a function to get the nearby venues in a given location. The function returns the geographical coordinates, the category of the venues in a given neighborhood together with its coordinates.

In [87]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

We call the function above and cast the output into a dataframe.

In [88]:
toronto_venues = getNearbyVenues(names=df_toronto['Neighborhood'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                  )
df = pd.DataFrame(toronto_venues)

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale, Willowdale East
York Mills West
Willowdale, Willowdale West
Parkwoods
Don Mills
Don Mills
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview
Downsview
Downsview
Downsview
Victoria Village
Parkview Hill, Woodbine Gardens
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto, Broadview North (Old East York)
The Danforth West, 

As we see below, there are 2127 venues in all the neighborhoods.

In [89]:
print(toronto_venues.shape)
toronto_venues.head()

(2127, 7)


Neighborhood  Neighborhood Latitude  \
0                          Malvern, Rouge              43.806686   
1  Rouge Hill, Port Union, Highland Creek              43.784535   
2  Rouge Hill, Port Union, Highland Creek              43.784535   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                   Venue  Venue Latitude  \
0              -79.194353                 Wendy’s       43.807448   
1              -79.160497  Chris Effects Painting       43.784343   
2              -79.160497   Royal Canadian Legion       43.782533   
3              -79.188711          RBC Royal Bank       43.766790   
4              -79.188711       G & G Electronics       43.765309   

   Venue Longitude              Venue Category  
0       -79.199056        Fast Food Restaurant  
1       -79.163742  Construction & Landscaping  
2       -79.163085                         Bar  
3       -79.191151                        Bank  
4       -79.191537           Electronics Store

Next, we group the venues by neighborhoods.

In [90]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            5   
Alderwood, Long Branch                                               6   
Bathurst Manor, Wilson Heights, Downsview North                     21   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   24   
...                                                                ...   
Willowdale, Willowdale East                                         34   
Willowdale, Willowdale West                                          6   
Woburn                                                               3   
Woodbine Heights                                                     6   
York Mills West                                                      2   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             5   
Alderwood, Long Branch                                                6   
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    24   
...                                                                 ...   
Willowdale, Willowdale East                                          34   
Willowdale, Willowdale West                                           6   
Woburn                                                                3   
Woodbine Heights                                                      6   
York Mills West                                                       2   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            5               5   
Alderwood, Long Branch                               6               6   
Bathurst Manor, Wilson Heights, Downsview North     21              21   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   24              24   
...                                                ...             ...   
Willowdale, Willowdale East                         34              34   
Willowdale, Willowdale West                          6               6   
Woburn                                               3               3   
Woodbine Heights                                     6               6   
York Mills West                                      2               2   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      5   
Alderwood, Long Branch                                         6   
Bathurst Manor, Wilson Heights, Downsview North               21   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             24   
...                                                          ...   
Willowdale, Willowdale East                                   34   
Willowdale, Willowdale West                                    6   
Woburn                                                         3   
Woodbine Heights                                               6   
York Mills West                                                2   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     5  
Alderwood, Long Branch                                        6 

We count the unique venue categories below:

In [187]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 268 uniques categories.


We will convert the venue categories to their one-hot vector representation.

In [188]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Train Station  \
0                 0                    0  ...              0   
1                 0                    0  ...              0   
2                 0                    0  ...              0   
3                 0                    0  ...              0   
4                 0                    0  ...              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                      0                0         0          0            0   
1                      0                0         0          0            0   
2                      0                0         0          0            0   
3                      0                0         0          0            0   
4                      0                0         0          0            0   

   Women's Store  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 268 columns]

In [189]:
toronto_onehot.shape

(2127, 268)

Next, we group the venues by neighborhood.

In [190]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Agincourt          0.0   
1                            Alderwood, Long Branch          0.0   
2   Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                   Bayview Village          0.0   
4                 Bedford Park, Lawrence Manor East          0.0   
..                                              ...          ...   
89                      Willowdale, Willowdale East          0.0   
90                      Willowdale, Willowdale West          0.0   
91                                           Woburn          0.0   
92                                 Woodbine Heights          0.0   
93                                  York Mills West          0.0   

    Accessories Store  Afghan Restaurant  Airport  Airport Food Court  \
0                 0.0                0.0      0.0                 0.0   
1                 0.0                0.0      0.0                 0.0   
2                 0.0                0.0      0.0                 0.0   
3                 0.0                0.0      0.0                 0.0   
4                 0.0                0.0      0.0                 0.0   
..                ...                ...      ...                 ...   
89                0.0                0.0      0.0                 0.0   
90                0.0                0.0      0.0                 0.0   
91                0.0                0.0      0.0                 0.0   
92                0.0                0.0      0.0                 0.0   
93                0.0                0.0      0.0                 0.0   

    Airport Gate  Airport Lounge  Airport Service  Airport Terminal  ...  \
0            0.0             0.0              0.0               0.0  ...   
1            0.0             0.0              0.0               0.0  ...   
2            0.0             0.0              0.0               0.0  ...   
3            0.0             0.0              0.0               0.0  ...   
4            0.0             0.0              0.0               0.0  ...   
..           ...             ...              ...               ...  ...   
89           0.0             0.0              0.0               0.0  ...   
90           0.0             0.0              0.0               0.0  ...   
91           0.0             0.0              0.0               0.0  ...   
92           0.0             0.0              0.0               0.0  ...   
93           0.0             0.0              0.0               0.0  ...   

    Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0             0.0                            0.0               0.0   
1             0.0                            0.0               0.0   
2             0.0                            0.0               0.0   
3             0.0                            0.0               0.0   
4             0.0                            0.0               0.0   
..            ...                            ...               ...   
89            0.0                            0.0               0.0   
90            0.0                            0.0               0.0   
91            0.0                            0.0               0.0   
92            0.0                            0.0               0.0   
93            0.0                            0.0               0.0   

    Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  \
0      0.000000               0.000000              0.0       0.0        0.0   
1      0.000000               0.000000              0.0       0.0        0.0   
2      0.000000               0.000000              0.0       0.0        0.0   
3      0.000000               0.000000              0.0       0.0        0.0   
4      0.000000               0.000000              0.0       0.0        0.0   
..          ...                    ...              ...       ...        ...   
89     0.000000               0

In [191]:
toronto_grouped.shape

(94, 268)

We then write a for loop to get the top 5 venues in each neighborhood in descending order.

In [192]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0             Breakfast Spot   0.2
1                     Lounge   0.2
2               Skating Rink   0.2
3             Clothing Store   0.2
4  Latin American Restaurant   0.2


----Alderwood, Long Branch----
            venue  freq
0     Pizza Place  0.33
1             Pub  0.17
2             Gym  0.17
3  Sandwich Place  0.17
4     Coffee Shop  0.17


----Bathurst Manor, Wilson Heights, Downsview North----
              venue  freq
0              Bank  0.10
1       Coffee Shop  0.10
2        Restaurant  0.05
3  Sushi Restaurant  0.05
4       Supermarket  0.05


----Bayview Village----
                 venue  freq
0  Japanese Restaurant  0.25
1                 Café  0.25
2                 Bank  0.25
3   Chinese Restaurant  0.25
4          Yoga Studio  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0          Restaurant  0.08
1      Sandwich Place  0.08
2  Italian Restaurant  0.08
3         Coffee Shop  0.

We can define a function to get the most common venues in a given neighborhood.

In [193]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Then, we will use the function above to write the most common 10 venues for each neighborhood. 

In [203]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0          Skating Rink  Latin American Restaurant        Clothing Store   
1                   Gym                Coffee Shop        Sandwich Place   
2           Coffee Shop              Shopping Mall           Pizza Place   
3                  Bank        Japanese Restaurant    Chinese Restaurant   
4        Sandwich Place         Italian Restaurant            Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot      Doner Restaurant    Dim Sum Restaurant   
1                   Pub         Women's Store        Discount Store   
2         Deli / Bodega        Ice Cream Shop   Fried Chicken Joint   
3         Women's Store    Dim Sum Restaurant                 Diner   
4       Thai Restaurant          Liquor Store             Juice Bar   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                      Diner        Discount Store    Distribution Center  
1           Department Store          Dessert Shop     Dim Sum Restaurant  
2  Middle Eastern Restaurant      Sushi Restaurant         Sandwich Place  
3             Discount Store   Distribution Center                Dog Run  
4                    Butcher                  Café      Indian Restaurant

We can use k-means clustering to group the neighborhoods into 6 categories. 

In [204]:
# set number of clusters
kclusters = 6

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

We will add the cluster category to our dataframe and merge it with the borough and geographical corrdinate information.

In [205]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head(50) # check the last columns!

Postal Code          Borough  \
0          M1B      Scarborough   
1          M1C      Scarborough   
2          M1E      Scarborough   
3          M1G      Scarborough   
4          M1H      Scarborough   
5          M1J      Scarborough   
6          M1K      Scarborough   
7          M1L      Scarborough   
8          M1M      Scarborough   
9          M1N      Scarborough   
10         M1P      Scarborough   
11         M1R      Scarborough   
12         M1S      Scarborough   
13         M1T      Scarborough   
14         M1V      Scarborough   
15         M1W      Scarborough   
16         M1X      Scarborough   
17         M2H       North York   
18         M2J       North York   
19         M2K       North York   
20         M2L       North York   
21         M2M       North York   
22         M2N       North York   
23         M2P       North York   
24         M2R       North York   
25         M3A       North York   
26         M3B       North York   
27         M3C       North York   
28         M3H       North York   
29         M3J       North York   
30         M3K       North York   
31         M3L       North York   
32         M3M       North York   
33         M3N       North York   
34         M4A       North York   
35         M4B        East York   
36         M4C        East York   
37         M4E     East Toronto   
38         M4G        East York   
39         M4H        East York   
40         M4J        East York   
41         M4K     East Toronto   
42         M4L     East Toronto   
43         M4M     East Toronto   
44         M4N  Central Toronto   
45         M4P  Central Toronto   
46         M4R  Central Toronto   
47         M4S  Central Toronto   
48         M4T  Central Toronto   
49         M4V  Central Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                                      Malvern, Rouge  43.806686 -79.194353   
1              Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497   
2                   Guildwood, Morningside, West Hill  43.763573 -79.188711   
3                                              Woburn  43.770992 -79.216917   
4                                           Cedarbrae  43.773136 -79.239476   
5                                 Scarborough Village  43.744734 -79.239476   
6         Kennedy Park, Ionview, East Birchmount Park  43.727929 -79.262029   
7                     Golden Mile, Clairlea, Oakridge  43.711112 -79.284577   
8     Cliffside, Cliffcrest, Scarborough Village West  43.716316 -79.239476   
9                         Birch Cliff, Cliffside West  43.692657 -79.264848   
10  Dorset Park, Wexford Heights, Scarborough Town...  43.757410 -79.273304   
11                                  Wexford, Maryvale  43.750072 -79.295849   
12                                          Agincourt  43.794200 -79.262029   
13            Clarks Corners, Tam O'Shanter, Sullivan  43.781638 -79.304302   
14  Milliken, Agincourt North, Steeles East, L'Amo...  43.815252 -79.284577   
15                      Steeles West, L'Amoreaux West  43.799525 -79.318389   
16                                        Upper Rouge  43.836125 -79.205636   
17                                  Hillcrest Village  43.803762 -79.363452   
18                       Fairview, Henry Farm, Oriole  43.778517 -79.346556   
19                                    Bayview Village  43.786947 -79.385975   
20                           York Mills, Silver Hills  43.757490 -79.374714   
21                            Willowdale, Newtonbrook  43.789053 -79.408493   
22                        Willowdale, Willowdale East  43.770120 -79.408493   
23                                    York Mills West  43.752758 -79.400049   
24                        Willowdale, Willowdale West  43.782736 -79.442259   
25                                          Parkwoods  43.753259 -79.329656   
26                                          Don Mills  43.745906 -79.352188   
27   

We then drop the rows that have NaN values for the venues.

In [206]:
toronto_merged.dropna(axis=0, how='any', inplace=True)

In [211]:
toronto_merged.head()


Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue  \
0 -79.194353             2.0  Fast Food Restaurant   
1 -79.160497             4.0                   Bar   
2 -79.188711             1.0     Electronics Store   
3 -79.216917             1.0           Coffee Shop   
4 -79.239476             1.0      Hakka Restaurant   

        2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0               Deli / Bodega           Event Space  Ethiopian Restaurant   
1  Construction & Landscaping         Women's Store      Doner Restaurant   
2                  Restaurant   Rental Car Location        Breakfast Spot   
3           Korean Restaurant            Donut Shop    Dim Sum Restaurant   
4                 Gas Station                  Bank                Bakery   

  5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0     Electronics Store  Eastern European Restaurant             Drugstore   
1    Dim Sum Restaurant                        Diner        Discount Store   
2          Intersection                         Bank        Medical Center   
3                 Diner               Discount Store   Distribution Center   
4   Fried Chicken Joint         Caribbean Restaurant       Thai Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Donut Shop      Doner Restaurant                Dog Run  
1   Distribution Center               Dog Run             Donut Shop  
2    Mexican Restaurant                 Diner     Dim Sum Restaurant  
3               Dog Run      Doner Restaurant          Women's Store  
4    Athletics & Sports        Cosmetics Shop             Comic Shop

We create a map to show the clusters of neighborhoods.

In [208]:
# create map
map_clusters = folium.Map(location=[43.7032, -79.3832], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters